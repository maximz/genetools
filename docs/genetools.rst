genetools package
=================

Submodules
----------

genetools.helpers module
------------------------

.. automodule:: genetools.helpers
   :members:
   :undoc-members:
   :show-inheritance:

genetools.palette module
------------------------

.. automodule:: genetools.palette
   :members:
   :undoc-members:
   :show-inheritance:

genetools.plots module
----------------------

.. automodule:: genetools.plots
   :members:
   :undoc-members:
   :show-inheritance:

genetools.scanpy\_helpers module
--------------------------------

.. automodule:: genetools.scanpy_helpers
   :members:
   :undoc-members:
   :show-inheritance:

genetools.stats module
----------------------

.. automodule:: genetools.stats
   :members:
   :undoc-members:
   :show-inheritance:

Module contents
---------------

.. automodule:: genetools
   :members:
   :undoc-members:
   :show-inheritance:
