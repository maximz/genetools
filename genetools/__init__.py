"""Top-level package for genetools."""

__author__ = """Maxim Zaslavsky"""
__email__ = "maxim@maximz.com"
__version__ = "0.4.0"

# We want genetools.[submodule].[function] to be accessible simply by importing genetools, without having to import genetools.[submodule]
from . import helpers, plots, scanpy_helpers, stats, trajectory

__all__ = ["helpers", "plots", "scanpy_helpers", "stats", "trajectory"]
