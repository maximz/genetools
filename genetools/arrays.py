from typing import Any, Optional, Union, List

import numpy as np
import pandas as pd


def get_top_n_percent(df: pd.DataFrame, col: str, fraction: float) -> pd.DataFrame:
    """Get top fraction `n` of a dataframe `df` by a specific column `col`"""
    return df.sort_values(col, ascending=False).head(n=int(fraction * df.shape[0]))


def get_top_n(df: pd.DataFrame, col: str, n: int) -> pd.DataFrame:
    """Get top amount `n` of a dataframe `df` by a specific column `col`"""
    return df.sort_values(col, ascending=False).head(n=n)


def weighted_mode(
    arr: Union[list, np.ndarray, pd.Series],
    weights: Union[List[int], np.ndarray, pd.Series],
) -> Any:
    """
    Get weighted mode (most common value) in array.
    Faster than sklearn.utils.extmath.weighted_mode but does not support axis vectorization.
    """
    return (
        pd.DataFrame({"key": arr, "value": weights})
        .groupby("key", observed=True, sort=False)["value"]
        .sum()
        .idxmax()
    )


def weighted_value_counts(
    df: pd.DataFrame,
    category_column_name: str,
    weight_column_name: str,
    normalize: bool = False,
    **groupby_kwargs,
) -> pd.Series:
    """Weighted value counts. Basically a sum of weight_column_name within each category_column_name group.

    If normalize is True (default False), the returned value counts will sum to 1.

    The optional groupby_kwargs are passed to the groupby procedure.
    For example, if category_column_name is a Categorical, passing observed=True will make sure that any unused categories are not included in the value counts.
    """
    counts = (
        df.groupby(category_column_name, **groupby_kwargs)[weight_column_name]
        .sum()
        .rename(category_column_name)
    )
    if normalize:
        return counts / counts.sum()
    return counts


def groupby_apply_weighted_value_counts(
    df: pd.DataFrame,
    *groupby_args,
    category_column_name: str,
    weight_column_name: str,
    normalize: bool = False,
    **groupby_kwargs,
) -> pd.Series:
    """
    This is to be used instead of:

    .. code-block:: python

        df.groupby(["columnA", "columnB"], observed=True)
            .apply(
                lambda grp: malid.external.genetools_arrays.weighted_value_counts(
                    grp,
                    "category_column",
                    "weight_column",
                    normalize=True,
                )
            )

    The preferred call is:

    .. code-block:: python

        genetools.arrays.groupby_apply_weighted_value_counts(
            df,
            ["columnA", "columnB"],
            observed=True,
            category_column_name="category_column",
            weight_column_name="weight_column",
            normalize=True
        )

    It's the same behavior with extra checks to make sure the output format has the expected shape: it should be a Series, like in standard groupby-value_counts.

    (Sometimes Pandas returns a DataFrame instead of a Series.)
    """
    result = df.groupby(*groupby_args, **groupby_kwargs).apply(
        lambda grp: weighted_value_counts(
            grp,
            category_column_name=category_column_name,
            weight_column_name=weight_column_name,
            normalize=normalize,
        )
    )
    if type(result) is pd.DataFrame:
        # special handling for DataFrame output edge case
        return result.stack()
    return result


def weighted_median(values: np.ndarray, weights: np.ndarray) -> Union[int, float]:
    """Weighted median: factor in the weights when finding center of the array."""
    # See also https://stackoverflow.com/a/35349142/130164 and https://stackoverflow.com/a/73905572/130164
    # Consider adding interpolation as in the second link?

    # defensive cast
    values = np.array(values)
    weights = np.array(weights)

    # sanity check
    if values.shape[0] != weights.shape[0]:
        raise ValueError("Values and weights should have the same shape.")

    # equivalent pandas code provided in comments to make this numpy version easier to understand:

    value_sort_order = np.argsort(values)
    # values_sorted = values.sort_values()

    weights_cumsum_in_value_sorted_order = np.cumsum(weights[value_sort_order])
    # weights_cumsum_in_value_sorted_order = weights.loc[values_sorted.index].cumsum()

    weights_sum = weights_cumsum_in_value_sorted_order[-1]  # same as weights.sum()
    weights_median_threshold = weights_sum / 2.0
    # weights_sum = weights_cumsum_in_value_sorted_order.iloc[-1] # same as weights.sum()
    # weights_median_threshold = weights_sum / 2.0

    # get threshold position
    threshold_position = np.searchsorted(
        weights_cumsum_in_value_sorted_order, weights_median_threshold
    )

    # get value at the threshold position
    return values[value_sort_order[threshold_position]]
    # return values_sorted.loc[weights_cumsum_in_value_sorted_order >= weights_median_threshold].iloc[0]


def strings_to_character_arrays(
    strs: Union[np.ndarray, List[str], pd.Series], validate_equal_lengths: bool = True
) -> np.ndarray:
    """Create character matrix by "viewing" strings as 1-character string arrays, then reshaping"""
    char_matrix = np.array(strs)
    char_matrix = (
        char_matrix.astype("bytes").view("S1").reshape((char_matrix.shape[0], -1))
    )

    if validate_equal_lengths and b"" in char_matrix:
        # Note: this won't trip up on spaces in the string, because those are " " and ASCII code 32, not "" and int code 0.
        raise ValueError("Input strings must be of equal lengths.")

    return char_matrix


def strings_to_numeric_vectors(
    strs: Union[np.ndarray, List[str], pd.Series], validate_equal_lengths: bool = True
) -> np.ndarray:
    """Convert strings to numeric vectors (one entry per character)"""
    numeric_arr = strings_to_character_arrays(
        strs, validate_equal_lengths=validate_equal_lengths
    ).view(np.uint8)

    if 0 in numeric_arr:
        # Replace blanks (0s -- from \x00) with np.nan. Cast to float first to support np.nan
        numeric_arr = numeric_arr.astype(float)
        np.place(numeric_arr, numeric_arr == 0.0, np.nan)

    return numeric_arr


def numeric_vectors_to_character_arrays(arr: np.ndarray) -> np.ndarray:
    """Reverse operation of strings_to_numeric_vectors"""
    return np.array(arr, dtype=np.uint8).view("c")


def make_consensus_vector(
    matrix: np.ndarray, frequencies: Union[np.ndarray, List[int], pd.Series]
) -> np.ndarray:
    """
    Get weighted mode for each position across a set of equal-length vectors.
    """
    # weighted mode at each position -> centroid
    # apply function to every column vector
    consensus_elements = np.apply_along_axis(
        lambda col: weighted_mode(col, frequencies), 0, matrix
    )
    return consensus_elements


def make_consensus_sequence(
    sequences: Union[np.ndarray, pd.Series, List[str]],
    frequencies: Union[np.ndarray, pd.Series, List[int]],
) -> str:
    """
    Get weighted mode for each character across a set of equal-length input strings.
    """
    sequences = np.array(sequences)  # defensive cast
    if sequences.shape[0] == 1:
        # If only one sequence, just return it
        # The rest of the code is legal for this case, but probably faster to short circuit
        return sequences.item(0)

    char_matrix = strings_to_character_arrays(sequences, validate_equal_lengths=True)

    # weighted mode for each character
    consensus_elements = make_consensus_vector(char_matrix, frequencies)

    # assemble consensus string from individual characters
    return "".join(consensus_elements.astype(str))


def _masked_vector_fill_for_argmin_argmax_output(
    masked_arr: np.ma.MaskedArray,
    output: Union[np.ndarray, float, int],
    axis: Optional[int] = None,
) -> Union[np.ndarray, float, int]:
    """Finalize our masked_argmin/masked_argmax output by correcting one piece:
    For rows/columns that are all maksed (all NaN), argmin/argmax will return 0. That's incorrect behavior.
    Set entries for all-NaN rows or columns to NaN.
    """
    # if axis None, we will have scalar output from argmin/argmax
    if np.isscalar(output):
        # if all are masked, output will be 0, which is incorrect
        # return nan instead
        return np.nan if masked_arr.mask.all(axis=axis) else output

    # for rows/columns that are all masked (all nan), argmin will return 0. that's incorrect behavior.
    # set entries for all-nan rows or columns to NaN
    # first, cast to float so we can later set some entries to np.nan, which is a float
    output = output.astype(float)
    # .any will return
    output[masked_arr.mask.all(axis=axis)] = np.nan

    return output


def masked_argmin(
    masked_arr: np.ma.MaskedArray, axis: Optional[int] = None
) -> Union[np.ndarray, float, int]:
    """
    argmin on masked array. return nan for row/column (depending on axis setting) of all nans
    """
    # run standard argmin
    argmin_output = masked_arr.argmin(
        fill_value=np.ma.minimum_fill_value(masked_arr), axis=axis
    )

    return _masked_vector_fill_for_argmin_argmax_output(
        masked_arr, argmin_output, axis=axis
    )


def masked_argmax(
    masked_arr: np.ma.MaskedArray, axis: Optional[int] = None
) -> Union[np.ndarray, float, int]:
    """
    argmax on masked array. return nan for row/column (depending on axis setting) of all nans
    """
    # run standard argmax
    argmax_output = masked_arr.argmax(
        fill_value=np.ma.maximum_fill_value(masked_arr), axis=axis
    )

    return _masked_vector_fill_for_argmin_argmax_output(
        masked_arr, argmax_output, axis=axis
    )


def convert_matrix_to_one_element_per_row(arr: np.ndarray) -> pd.DataFrame:
    """
    record each element of 2d matrix as one entry in dataframe, with the row and column ids stored as well as the value
    """
    # remove any index or column names
    df = pd.DataFrame(np.array(arr))

    # melt
    df = (
        df.rename_axis("row_id", axis="index")  # rename index
        .reset_index()  # index is now a column called row_id
        .melt(id_vars="row_id", var_name="col_id", value_name="value")
    )

    # cast from object to int
    df["col_id"] = df["col_id"].astype(int)

    return df


def get_trim_both_sides_mask(
    a: Union[np.ndarray, pd.DataFrame], proportiontocut: float, axis: int = 0
) -> np.ndarray:
    """returns mask that applies consistent trim-both-sides learned on one array.

    suppose you have a data array and a weights array. you want to trimboth() the data array but keep the element weights aligned.

    solution:

    .. code-block:: python

        trimming_mask = genetools.arrays.get_trim_both_sides_mask(data, proportiontocut=0.1)
        return data[trimming_mask], weights[trimming_mask]

    """
    # based on scipy.stats.trimboth and scipy.stats.trim_mean
    a = np.asarray(a)

    if a.size == 0:
        return slice(None)

    if axis is None:
        a = a.ravel()
        axis = 0

    nobs = a.shape[axis]
    lowercut = int(proportiontocut * nobs)
    uppercut = nobs - lowercut
    if lowercut >= uppercut:
        raise ValueError("Proportion too big.")

    # np.partition creates a copy of the array with its elements rearranged in such a way that the value of the element in k-th position is in the position it would be in a sorted array.
    # All elements smaller than the k-th element are moved before this element and all equal or greater are moved behind it.
    # The ordering of the elements in the two partitions is undefined.

    # tl,dr: partial sort so that anything < lowercut-th index is before it in the array, and anything >= uppercut-th index is after it in the array.
    # that way we can just chop off the two ends

    # but instead let's get the indices that partition() would keep - this is argpartition()
    # note: often used with take_along_axis() to apply the indices.
    # these are equivalent: np.take_along_axis(x, np.argpartition(x, kth=1, axis=-1), axis=-1) == np.partition(x, kth=1)

    index_array = np.argpartition(a=a, kth=(lowercut, uppercut - 1), axis=axis)

    # generate an empty slicer along each axis of the array
    ndim = index_array.ndim  # per original scipy: should this be np.take_along_axis(a, index_array, axis=axis).ndim or a.ndim instead?
    sl = [slice(None)] * ndim
    # along the axis that matters, apply a mask that will cut the ends off
    sl[axis] = slice(lowercut, uppercut)
    mask = tuple(sl)

    return index_array[mask]


def make_dummy_variables_in_specific_order(
    values: Union[pd.Series, List[str]],
    expected_list: List[str],
    allow_missing_entries: bool,
) -> pd.DataFrame:
    """
    Create dummy variables in a defined order. All the values are confirmed to be in the "expected order" list.
    If an entry from "expected order" list is not present, still include as a dummy variable with all 0s if allow_missing_entires is True, or throw error otherwise.
    """
    # Defensive cast to Series,
    # and remove any unused categories in case it's a Categorical
    values = pd.Series(values).astype("category").cat.remove_unused_categories()

    if values.isna().any():
        raise ValueError("Some input entries are blank")

    dummy_vars = pd.get_dummies(values)

    # If values was a Categorical series, we won't be able to add new columns below to dummy_vars.
    # So cast the columns to string.
    dummy_vars.columns = dummy_vars.columns.astype(str)

    for value in expected_list:
        # Check whether each expected value is present in this dataset.
        if value not in dummy_vars.columns:
            if not allow_missing_entries:
                raise ValueError(
                    f"Entry {value} was expected, but was not found in the dataset."
                )
            # insert new column of all 0s (never present)
            dummy_vars[value] = 0

    if set(dummy_vars.columns) != set(expected_list):
        raise ValueError(
            "There are some values included that are not in the list of expected/possible values."
        )

    # Reorder
    dummy_vars = dummy_vars[expected_list]

    return dummy_vars
