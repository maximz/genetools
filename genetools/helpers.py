"""Pandas/Numpy common recipes."""

import os
import scipy
import numpy as np
import pandas as pd


def rename_duplicates(series, delim="-"):
    """Rename duplicate values to be unique. ['a', 'a'] will become ['a', 'a-1'], for example.

    :param series: series with values to rename
    :type series: pandas.Series
    :param delim: delimeter before duplicate-number index, defaults to "-"
    :type delim: str, optional
    :return: series where original duplicates have been renamed to -1, -2, etc.
    :rtype: pandas.Series
    """
    duplicate_suffix = (
        series.groupby(series).cumcount().astype(str).replace("0", "")
    )  # a number for all but first occurence
    extra_strs = delim + duplicate_suffix
    # remove entries that are just the delim
    extra_strs = extra_strs.replace(delim, "")
    # add to values
    out = series.astype(str) + extra_strs
    # confirm unique (may fail if a-1 happened to match another element that preexisted!)
    assert out.nunique() == out.shape[0]
    return out


def merge_into_left(left, right, **kwargs):
    """Defensively merge [right] series or dataframe into [left] by index, preserving [left]'s index exactly. [right] data will be reordered to match [left] index.

    :param left: left data whose index will be preserved
    :type left: pandas.DataFrame or pandas.Series
    :param right: right data which will be reordered based on left index.
    :type right: pandas.DataFrame or pandas.Series
    :param \**kwargs: passed to pandas.merge
    :return: left-merged DataFrame with [left]'s index
    :rtype: pandas.DataFrame
    """
    # defensively cast to dataframe
    df1 = pd.DataFrame(left)
    df2 = pd.DataFrame(right)
    df = pd.merge(
        df1,
        df2,
        how="left",
        left_index=True,
        right_index=True,
        sort=False,
        validate="1:1",
        **kwargs
    )
    # TODO: asserts are stripped away when code is optimized; replace with if not, raise ValueError('message')
    assert df.shape[0] == df1.shape[0]
    assert df.shape[1] == df1.shape[1] + df2.shape[1]
    df.index = df1.index
    return df


def sample_cells_from_clusters(obs_df, n_cells, cluster_key, cluster_names):
    """[summary]

    :param obs_df: [description]
    :type obs_df: [type]
    :param n_cells: [description]
    :type n_cells: [type]
    :param cluster_key: [description]
    :type cluster_key: [type]
    :param cluster_names: [description]
    :type cluster_names: [type]
    :return: [description]
    :rtype: [type]
    """
    return (
        obs_df[obs_df[cluster_key].isin(cluster_names)]
        .index.to_series()
        .sample(n_cells)
        .values
    )


def horizontal_concat(df_left, df_right):
    """Concatenate df_right horizontally to df_left, with no checks for whether the indexes match, but confirming final shape.

    :param df_left: Left data
    :type df_left: pandas.DataFrame or pandas.Series
    :param df_right: Right data
    :type df_right: pandas.DataFrame or pandas.Series
    :return: Copied dataframe with df_right's columns glued onto the right side of df_left's columns
    :rtype: pandas.DataFrame
    """
    # defensively cast to DataFrame
    df1 = pd.DataFrame(df_left)
    df2 = pd.DataFrame(df_right)

    df = pd.concat([df1, df2], axis=1)
    assert df.shape[0] == df1.shape[0] == df2.shape[0]
    assert df.shape[1] == df1.shape[1] + df2.shape[1]
    return df


def vertical_concat(df_top, df_bottom, reset_index=False):
    """Concatenate df_bottom vertically to df_top, with no checks for whether the columns match, but confirming final shape.

    :param df_top: Top data
    :type df_top: pandas.DataFrame
    :param df_bottom: Bottom data
    :type df_bottom: pandas.DataFrame
    :param reset_index: Reset index values after concat, defaults to False
    :type reset_index: bool, optional
    :return: Copied dataframe with df_bottom's rows glued onto the bottom of df_top's rows
    :rtype: pandas.DataFrame
    """
    # defensively cast to DataFrame
    df1 = pd.DataFrame(df_top)
    df2 = pd.DataFrame(df_bottom)

    df = pd.concat([df1, df2], axis=0)
    if reset_index:
        # so far indexes have just been glued together
        # we can reset it to be unique values
        df = df.reset_index(drop=True)
    assert df.shape[0] == df1.shape[0] + df2.shape[0]
    assert df.shape[1] == df1.shape[1] == df2.shape[1]
    return df


def barcode_split(
    obs_names, separator="-", colname_barcode="barcode", colname_library="library_id"
):
    """Split single cell barcodes such as ATGC-1 into a barcode column with value "ATGC" and a library ID column with value 1.

    Recommended usage with scanpy:
    adata.obs = horizontal_concat(adata.obs, barcode_split(adata.obs_names))

    :param obs_names: Cell barcodes with a library ID suffix.
    :type obs_names: pandas.Series or pandas.Index
    :param separator: library ID separator, defaults to '-'
    :type separator: str, optional
    :param colname_barcode: output column name containing barcode without library ID suffix, defaults to 'barcode'
    :type colname_barcode: str, optional
    :param colname_library: output column name containing library ID suffix as an int, defaults to 'library_id'
    :type colname_library: str, optional
    :return: Two-column dataframe containing barcode prefix and library ID suffix.
    :rtype: pandas.DataFrame
    """
    # defensively cast to a string Series in case an Index was passed, such as adata.obs_names or adata.obs.index
    df = pd.Series(obs_names, dtype="str").str.split(separator, expand=True)
    df.columns = [colname_barcode, colname_library]
    df[colname_library] = df[colname_library].astype(int)
    return df


def get_off_diagonal_values(arr):
    """Get off-diagonal values of a numpy 2d array as a flattened 1d array.

    :param arr: input numpy 2d array
    :type arr: numpy.ndarray
    :return: flattened 1d array of non-diagonal values only
    :rtype: numpy.ndarray
    """
    # See https://stackoverflow.com/a/35746928/130164
    return arr[~np.eye(arr.shape[0], dtype=bool)].flatten()


def make_slurm_command(
    script,
    job_name,
    log_path,
    env=None,
    options={},
    job_group_name="",
    wrap_script=True,
):
    """Generate slurm sbatch command. Should be pipe-able straight to bash.

    Automatic log filenames will take the format:
        - `{{ log_path }}/{{ job_group_name (optional) }}/{{ job_name }}.out` for stdout
        - `{{ log_path }}/{{ job_group_name (optional) }}/{{ job_name }}.err` for stderr

    You can override automatic log filenames by manually supplying "output" and "error" values in the `options` dict.

    :param script: path to an executable script, or inline script (if wrap_script is True)
    :type script: str
    :param job_name: job name, used for naming log files
    :type job_name: str
    :param log_path: destination for log files.
    :type log_path: str
    :param env: any environment variables to pass to script, defaults to None
    :type env: dict, optional
    :param options: any CLI options for sbatch, defaults to {}
    :type options: dict, optional
    :param job_group_name: optional group name for this job and related jobs, used for naming log files, defaults to ""
    :type job_group_name: str, optional
    :param wrap_script: whether the script is inline as opposed to a file on disk, defaults to True
    :type wrap_script: bool, optional
    :return: an sbatch command
    :rtype: str
    """
    log_fname_prefix = os.path.join(log_path, job_group_name, job_name)
    if "output" not in options:
        options["output"] = log_fname_prefix + ".out"
    if "error" not in options:
        options["error"] = log_fname_prefix + ".err"

    options_items = ['--%s="%s"' % (name, val) for name, val in options.items()]
    options_string = " ".join(options_items)

    variable_string = ""
    if env is not None:
        variable_items = ['"%s"="%s"' % (name, val) for name, val in env.items()]
        variable_string = "--export=" + ",".join(variable_items)

    # Very important to wrap the "wrap script" in single quotes,
    # so that the script will pick up the exported variables during execution.
    script_string = "--wrap='%s'" % script if wrap_script else script
    return "sbatch {options} {variables} {script};".format(
        options=options_string, variables=variable_string, script=script_string
    )
