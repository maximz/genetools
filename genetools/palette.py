import collections
import dataclasses
import matplotlib as mpl


def convert_palette_list_to_dict(palette, hue_names, sort_hues=True):
    """
    If palette is a list, convert it to a dict, assigning a color to each value in hue_names (with sort enabled by default).

    If palette is already a dict, pass it through with no changes.
    """
    if isinstance(palette, collections.abc.Mapping):
        # already a dict. do nothing.
        return palette
    if len(palette) < len(hue_names):
        raise ValueError(
            f"Not enough colors in palette. palette has {len(palette)} colors, but hue_names has {len(hue_names)} values."
        )
    if sort_hues:
        hue_names = sorted(hue_names, key=lambda hue: str(hue))
    return {k: v for k, v in zip(hue_names, palette)}


@dataclasses.dataclass
class HueValueStyle:
    """
    Describes how to style a particular value (category) of a categorical hue column.

    Use palettes mapping hue values to HueValueStyles to make plots with different marker shapes, transparencies, z-orders, etc. for different groups.

    The plotting functions accept a hue_key, which identifies a dataframe column that contains hue values.
    They also accept a palette mapping each hue value to a HueValueStyle that defines not just the color to use for that hue value, but also other styles:

    * Scatterplot marker shape, primary color, face color, edge color, line width, transparency, and line width.
    * Rectangle/barplot color and hatch pattern.
    * Size scale factor for scatterplot markers and legend entries. (The palette of HueValueStyles is defined separately from choosing marker size, and can be plotted at any selected base marker size.)

    Here's an example of assigning a custom HueValueStyle to a hue value in a color palette.
    This defines a custom unfilled shape, a custom z-order, and more:

    .. code-block:: python

        palette = {
            "group_A": genetools.palette.HueValueStyle(
                color=sns.color_palette("bright")[0],
                edgecolors=sns.color_palette("bright")[0],
                facecolors="none",
                marker="^",
                marker_size_scale_factor=1.5,
                linewidths=1.5,
                zorder=10,
            ),
            ...
        }

    For face and edge colors, ``None`` is the default value; to disable them, set to string ``'none'``.
    """

    color: str
    marker: str = None
    marker_size_scale_factor: float = 1.0
    legend_size_scale_factor: float = 1.0
    facecolors: str = None
    edgecolors: str = None
    linewidths: float = None
    zorder: int = 1
    alpha: float = None
    hatch: str = None

    @staticmethod
    def huestyles_to_colors_dict(d: dict) -> dict:
        """Cast any HueValueStyle values in dict to be color strings."""
        return {k: v.color if isinstance(v, HueValueStyle) else v for k, v in d.items()}

    @classmethod
    def from_color(cls, s):
        """Construct from color string only; keep all other marker parameters set to defaults.
        If already a HueValueStyle, pass through without modification."""
        if isinstance(s, cls):
            return s
        return cls(color=s)

    def apply_defaults(self, defaults: "HueValueStyle"):
        """
        Returns new HueValueStyle that applies defaults:
        Modifies this style to fill any missing values with the values from another HueValueStyle.

        Use case: supply global style defaults for an entire scatterplot, then override with customizations in any individual hue value style.
        """
        # The type hint is quoted because inside of the HueValuStyle class, the class itself is not defined yet. https://stackoverflow.com/a/44798831/130164

        # we are detecting unset values by checking whether they are None
        # to make this more robust, we should compare to each field's default value or default factory
        # see https://stackoverflow.com/a/56452356/130164
        changes = {
            field.name: getattr(defaults, field.name)
            for field in dataclasses.fields(defaults)
            if getattr(self, field.name) is None
        }

        # returns copy
        return dataclasses.replace(self, **changes)

    @staticmethod
    def _get_default_marker_size():
        """Return default marker size used by ax.scatter."""
        return mpl.rcParams["lines.markersize"] ** 2

    def render_scatter_props(self, marker_size=None):
        """Returns kwargs to pass to ax.scatter() to apply this style."""
        if marker_size is None:
            marker_size = self._get_default_marker_size()

        return dict(
            color=self.color,
            marker=self.marker,
            s=marker_size * self.marker_size_scale_factor,
            facecolors=self.facecolors,
            edgecolors=self.edgecolors,
            linewidths=self.linewidths,
            zorder=self.zorder,
            alpha=self.alpha,
        )

    def render_scatter_continuous_props(self, marker_size=None):
        """Returns kwargs to pass to ax.scatter() to apply this style, in the context of continuous cmap scatterplots."""
        if marker_size is None:
            marker_size = self._get_default_marker_size()
        return dict(
            marker=self.marker,
            s=marker_size * self.marker_size_scale_factor,
            edgecolors=self.edgecolors,
            linewidths=self.linewidths,
            zorder=self.zorder,
            alpha=self.alpha,
        )

    def render_scatter_legend_props(self):
        """Returns kwargs to pass to ax.legend() to apply this style."""
        # Apply scaling to the default marker size we'd get by calling ax.scatter without any size arguments.
        legend_marker_size = (
            self._get_default_marker_size() * self.legend_size_scale_factor
        )

        return dict(
            color=self.color,
            marker=self.marker,
            s=legend_marker_size,
            facecolors=self.facecolors,
            edgecolors=self.edgecolors,
            linewidths=self.linewidths,
        )

    def render_rectangle_props(self):
        """Returns kwargs to pass to ax.bar() to apply this style."""
        return dict(color=self.color, hatch=self.hatch)
