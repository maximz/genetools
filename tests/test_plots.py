#!/usr/bin/env python

"""Tests for `genetools` plots.

Keep in mind when writing plotting tests:
- Use `@pytest.mark.mpl_image_compare` decorator to automatically do snapshot testing. See README.md for how to regenerate snapshots.
- `plt.tight_layout()` seems to produce different figure dimensions across different platforms.
    - Avoid until we figure out how to generate figures in a consistent way with Travis CI. (Run Travis-like environment locally in Docker? Hacky.)
- Some figures seem not to export right unless you save with tight bounding box:
    - `@pytest.mark.mpl_image_compare(savefig_kwargs={"bbox_inches": "tight"})`
"""

import pytest
import numpy as np
import pandas as pd
import random
import matplotlib

matplotlib.use("Agg")

from genetools import plots

random_seed = 12345
np.random.seed(random_seed)
random.seed(random_seed)


@pytest.mark.mpl_image_compare(savefig_kwargs={"bbox_inches": "tight"})
def test_umap_scatter_discrete(adata):
    """Test umap_scatter with discrete hue."""
    fig, _ = plots.umap_scatter(
        data=adata.obs,
        umap_1_key="umap_1",
        umap_2_key="umap_2",
        hue_key="louvain",
        label_key="louvain",
    )
    return fig


@pytest.mark.mpl_image_compare
def test_umap_scatter_continuous(adata):
    """Test umap_scatter with continouous hue."""
    fig, _ = plots.umap_scatter(
        data=adata.obs,
        umap_1_key="umap_1",
        umap_2_key="umap_2",
        hue_key="CST3",
        continuous_hue=True,
        label_key="louvain",
    )
    return fig


@pytest.mark.mpl_image_compare(savefig_kwargs={"bbox_inches": "tight"})
def test_horizontal_stacked_bar_plot():
    df = pd.DataFrame(
        {
            "cluster": [
                "Cluster 1",
                "Cluster 1",
                "Cluster 2",
                "Cluster 2",
                "Cluster 3",
                "Cluster 3",
            ],
            "cell_type": ["B cell", "T cell", "B cell", "T cell", "B cell", "T cell"],
            "frequency": [0.25, 0.75, 15, 5, 250, 750],
        }
    )
    fig, _ = plots.horizontal_stacked_bar_plot(
        df,
        index_key="cluster",
        hue_key="cell_type",
        value_key="frequency",
        normalize=True,
    )
    return fig
