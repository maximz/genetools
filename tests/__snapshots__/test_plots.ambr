# serializer version: 1
# name: test_pdf_deterministic_output
  '46fd883223328ffec5a38ae6433975b7'
# ---
